cli/cmd/bb/bb.go

```diff
@@ -51,72 +55,139 @@ func main() {
 	os.Exit(exitCode)
 }
 
-func run() (exitCode int, err error) {
-	start := time.Now()
-	// Record original arguments so we can show them in the UI.
-	originalArgs := append([]string{}, os.Args...)
-
-	args := handleGlobalCliFlags(os.Args[1:])
-
+// StartDebug prints relevant debug information related to CLI startup. It is
+// intended to be called after Configure, as Configure sets log levels.
+func StartupDebug(start time.Time) {
 	log.Debugf("CLI started at %s", start)
 	log.Debugf("args[0]: %s", os.Args[0])
 	log.Debugf("env: BAZEL_REAL=%s", os.Getenv("BAZEL_REAL"))
 	log.Debugf("env: BAZELISK_SKIP_WRAPPER=%s", os.Getenv("BAZELISK_SKIP_WRAPPER"))
 	log.Debugf("env: USE_BAZEL_VERSION=%s", os.Getenv("USE_BAZEL_VERSION"))
 	log.Debugf("env: BB_USE_BAZEL_VERSION=%s", os.Getenv("BB_USE_BAZEL_VERSION"))
+}
 
-	if _, err := bazelisk.ResolveVersion(); err != nil {
-		log.Printf("Failed to resolve bazel version: %s", err)
+// Configure takes a slice of options and filters them for the global options
+// needed to configure logging and the watcher, accumulates those options to get
+// the necessary values, and then configures logging and the watcher.
+//
+// We template this top support different types of Option slices; for example,
+// []*parsed.IndexedOption
+func Configure[O options.Option](bbOpts []O) {
```

Comment:
uber nit: can we use `T` here instead of `O`? The `O` looks like a `0` :sweat_smile: